Add async-backup

contrapunctus-1 opened the pull request:
### Brief summary of what the package does

Provides a function to make backups of the current file asynchronously.

### Direct link to the package repository

https://tildegit.org/contrapunctus/async-backup

### Your association with the package

Author and maintainer

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

riscy commented:
Thanks, looks like a good idea.  Some quick feedback:

- Could `(concat (locate-user-emacs-file "" "/async-backup"))` just be `(locate-user-emacs-file "async-backup")`?
- I may be missing something with respect to cross-platform compatibility, but is there a reason to fire up another `emacs` process instead of invoking another builtin like `cp`?
- Instead of creating a buffer called `"*async-backup*"`, could you set that variable to nil and use `set-process-sentinel` to redirect the output to `*Messages*`?  This might make things a little more seamless
- Lastly, just thinking as I type, but I wonder if another way to package this up would be to create a minor mode that advises the function `backup-buffer` -- and then you would implicitly be able to make use of `make-backup-files` and the other related variables, if that makes sense

contrapunctus-1 commented:
> Could `(concat (locate-user-emacs-file "" "/async-backup"))` just be `(locate-user-emacs-file "async-backup")`?

Good catch, thanks. I've updated it.

> I may be missing something with respect to cross-platform compatibility, but is there a reason to fire up another emacs process instead of invoking another builtin like cp?

`cp` is what I used in my first prototype - I changed it because I figured it wouldn't work on non-UNIX systems. Haven't tested either on anything except a Linux system, though...

> Instead of creating a buffer called `"*async-backup*"`, could you set that variable to nil and use `set-process-sentinel` to redirect the output to `*Messages*`? This might make things a little more seamless

Hm...at the moment it just says "Process async-backup finished" - does it really help to have it in the Messages buffer? 🤔

> Lastly, just thinking as I type, but I wonder if another way to package this up would be to create a minor mode that advises the function `backup-buffer` -- and then you would implicitly be able to make use of `make-backup-files` and the other related variables, if that makes sense

An interesting idea, thanks. Perhaps I could make it an addition to this package, so users can choose to use `async-backup` or `async-backup-minor-mode` - the former being the functionality as it exists, and the latter advising the existing backup function(s) such that the built-in backup configuration is used (and `async-backup` 's custom variables ignored).

riscy commented:
> does it really help to have it in the Messages buffer?

That's a fair question - I guess I bring it up because if many packages left incidental logs in different buffers, users could end up in a situation with lots of extra buffers floating around.  I don't think this should be a blocker though, so I'll merge now.
